validate.sh

```diff
@@ -0,0 +1,183 @@
+
+declare -a uobjects=("ccsynchbench.run"                     "dsmsynchbench.run" "hsynchbench.run" "oscibench.run"      "simbench.run")
+declare -a queues=(  "ccqueuebench.run" "clhqueuebench.run" "dsmqueuebench.run" "hqueuebench.run" "osciqueuebench.run" "simqueuebench.run")
+declare -a stacks=(  "ccstackbench.run" "clhstackbench.run" "dsmstackbench.run" "hstackbench.run" "oscistackbench.run" "simstackbench.run")
+
+declare -a validate_objects=()
+
+validate_objects+=${uobjects[@]}
+validate_objects+=${queues[@]}
+validate_objects+=${stacks[@]}
+
+if [ "$1" = "--help" ] || [ "$1" = "-h" ]; then
+    usage;
+    exit;
+fi
+
+while [ "$1" != "" ]; do
+    PARAM=`echo $1 | awk -F= '{print $1}'`
+    VALUE=`echo $1 | awk -F= '{print $2}'`
+    case $PARAM in
+        -h | --help)
+            usage;
+            exit;
+            ;;
+        -s | --step_threads)
+            STEP_THREADS=$VALUE
+            STEP_SELETCTED=1
+            ;;
+        -t | --max_threads)
+            MAX_PTHREADS=$VALUE
+            ;;
+        -f | --fibers)
+            FIBERS="-f $VALUE"
+            ;;
+        -w | --max_work)
+            WORKLOAD="-w $VALUE"
+            ;;
+        -n | --numa_nodes)
+            NUMA_NODES="-n $VALUE"
+            ;;
+        -r | --runs)
+            RUNS_PER_THREAD=$VALUE
+            ;;
+        -*)
+            echo "ERROR: unknown parameter \"$PARAM\""
+            usage
+            exit 1
+            ;;
+	*)
+	    FILE=$PARAM
+	    ;;
+    esac
+    shift
+done
+
+# Calculate step according to user preferences
+if [ $STEP_SELETCTED -eq 0 ]; then
+    if [ $MAX_PTHREADS -gt 8 ]; then 
+        STEP_THREADS=$(($MAX_PTHREADS / 8))
+    fi
+fi
+
+# Fill the array with the selected number of threads
+PTHREADS_ARRAY=()
+
+# Add set of iterations for 1 thread, if needed
+if [ $STEP_THREADS -ne 1 ]; then
+    PTHREADS_ARRAY+=(1)
+fi
+# Add intermediate sets
+for (( PTHREADS=$STEP_THREADS; PTHREADS<$MAX_PTHREADS; PTHREADS+=$STEP_THREADS ));do
+    PTHREADS_ARRAY+=($PTHREADS)
+done
+# Add set of iterations for max threads
+PTHREADS_ARRAY+=($MAX_PTHREADS)
+
+echo "Compiling the sources..."
+make clean debug > $BUILD_LOG
+
+# Run the selected number of threads
+for PTHREADS in "${PTHREADS_ARRAY[@]}"; do
+    printf "\n\e[36mValidating for %3d threads\n" $PTHREADS
+    echo -e "==========================\e[39m"
+    for bench in "${uobjects[@]}"; do
+        printf "Validating %-20s \t\t\t\t\t" $bench
+        runs=$(($RUNS_PER_THREAD * $PTHREADS))
+        $BIN_PATH/$bench -t $PTHREADS -r $runs $WORKLOAD $FIBERS $NUMA_NODES > $RES_FILE 2>&1
+        state=$(fgrep "Object state: " $RES_FILE)
+        state=${state/#"DEBUG: Object state: "}
+        if [ $state -eq $runs ]; then
+            echo -e $COLOR_PASS
+        else
+            echo -e $COLOR_FAIL
+            echo "Expected state: " $runs
+            echo "Invalid state: " $state
+            PASS_STATUS=0
+        fi
+    done
+
+    for bench in "${stacks[@]}"; do
```

Comment:
In general I think that this loop and the previous one are exactly the same. Maybe you can merge them into one.
Less code to inspect and maintain.

Reply:
The problem here is that for different class of objects, the validation process is a bit different.
Yes, we could merge these loops to a big one, but we should insert a lot of if-conditions in the loop.
In this case, I think the code maybe a bit unreadable.